### DISCUSSION TITLE
Why so slow?

### DISCUSSION THREAD

**Discussion by ValerioB88** [mesa/mesa#1124] [Q&A]
this is not a critique of the tool, which I really appreciate, but a request of clarification. I have tried a "stripped down" version of mesa to see how fast it can get, and it's not as fast as I expected, and I would like to understand why.

In particular, I looked at the "boid_flockers" example. In the `server.py` I commented out the `boid_canvas` so there is no visualization in the server. 
Then in `model.py`, in  `BoidFlockers.step`, I comment out the `self.schedule.step`, so that the model doesn't do anything.

The maximum frames per second by default is 20, so I changed the js code to get the slider up to 1000 fps. 

I run the simulation and rump up the fps slider to 1000fps. However, 1000fps is never reached: I measure around 125fps.

As far as I understand, the code isn't doing much apart from network communication.  So is this value expected? Is the boid_flocking  example doing something more under the hood? (I don't see any data collection, but I am new to mesa so I might have missed it).
Is this "delay" completely due to websocket communication? Is it expected? Is there any way to speed it up?
Thanks

**Comment by rht** [comment]
(This is not an answer) I haven't spent much time on these part of the codebase, and so can't give you relevant info, but just wanting to encourage you that any opportunities for performance optimization are welcome!

Though in most medium-big models, it is likely that `self.schedule.step` is taking the bulk of the time. But you could quantify if e.g. the network communication, redraw, etc, is roughly 0.5x the step of a 20x20 Schelling model or something (I made up the numbers), on your hardware.

**Comment by tpike3** [comment]
@ValerioB88 We are having a dev meeting this [Saturday](https://github.com/projectmesa/mesa/discussions/1094). Our front end is always a topic of discussion and we have even talked about rebuilding it but have lacked what we feel is the requisite expertise. 

Your more than welcome to attend if interested.

**Comment by Corvince** [comment]
I looked a bit into this, because I remembered how much fun profiling I had in the past. Unfortunately the frame rate that your are seeing is just how fast (or slow) the model is running. I timed model runs with and without the visualization timer and they are the same.

**Reply by ValerioB88** [reply]
Thanks for checking, so all the server stuff is not taking much time at all. That's pretty reassuring to be honest, I woulnd't know how to optimize that side - but I have some hopes about the model itself.

**Reply by Corvince** [reply]
No problem. For high frame rates the frontend will of course have some impact, because it requests the next step only after it has finished drawing. Don't know your use case but I wouldn't worry about that for now.

Otherwise mesa is not a high-performance framework, but most models are rather strongly explicit and should be easy to follow. The biggest performance bottleneck is that each agents `step` function is called individually. So for 100 boids `agent.step` is called 100 times. You could probably leverage the numpy space to do the calculations at once (not using `agent.step`). This would probably increase the speed ~100x. But the model would look quite differently from usual mesa models (but you could still use things like data collection and the frontend)

**Reply by ValerioB88** [reply]
Yep, it makes sense. In fact, in my own [EvoAgent ](https://github.com/ValerioB88/EvoAgent)project, I do a preprocessing step computing all the distances between agents and food vectorially, which results in a massive speed up, breaking just a little bit the mesa framework. :)

**Comment by jackiekazil** [comment]
I believe this question has been answered, so I am going to close it for now for clean-up. 
I will also add we are consistently making improvements to Mesa and hopefully more of those will be coming this summer on the front end.
